In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("hw7.ipynb")

# CPSC 330 - Applied Machine Learning 

## Homework 7: Word embeddings and topic modeling 
**Due date: See the [Calendar](https://htmlpreview.github.io/?https://github.com/UBC-CS/cpsc330/blob/master/docs/calendar.html).**

## Imports

In [2]:
import os

%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline, make_pipeline

<br><br>

## Submission instructions
<hr>
rubric={points:2}

You will receive marks for correctly submitting this assignment. To submit this assignment, follow the instructions below:

- **You may work on this assignment in a group (group size <= 4) and submit your assignment as a group.** 
- Below are some instructions on working as a group.  
    - The maximum group size is 4. 
    - You can choose your own group members. 
    - Use group work as an opportunity to collaborate and learn new things from each other. 
    - Be respectful to each other and make sure you understand all the concepts in the assignment well. 
    - It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. [Here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members) are some instructions on adding group members in Gradescope.  
- Upload the .ipynb file to Gradescope.
- **If the .ipynb file is too big or doesn't render on Gradescope for some reason, also upload a pdf or html in addition to the .ipynb.** 
- Make sure that your plots/output are rendered properly in Gradescope.

<br><br><br><br>

## Exercise 1:  Exploring pre-trained word embeddings <a name="1"></a>
<hr>

In lecture 18, we talked about natural language processing (NLP). Using pre-trained word embeddings is very common in NLP. It has been shown that pre-trained word embeddings work well on a variety of text classification tasks. These embeddings are created by training a model like Word2Vec on a huge corpus of text such as a dump of Wikipedia or a dump of the web crawl. 

A number of pre-trained word embeddings are available out there. Some popular ones are: 

- [GloVe](https://nlp.stanford.edu/projects/glove/)
    * trained using [the GloVe algorithm](https://nlp.stanford.edu/pubs/glove.pdf) 
    * published by Stanford University 
- [fastText pre-trained embeddings for 294 languages](https://fasttext.cc/docs/en/pretrained-vectors.html) 
    * trained using the fastText algorithm
    * published by Facebook
    
In this exercise, you will be exploring GloVe Wikipedia pre-trained embeddings. The code below loads the word vectors trained on Wikipedia using an algorithm called Glove. You'll need `gensim` package in your cpsc330 conda environment to run the code below. 

```
> conda activate cpsc330
> conda install -c anaconda gensim
```

In [3]:
import sys
!{sys.executable} -m pip install gensim

In [4]:
import gensim
import gensim.downloader

print(list(gensim.downloader.info()["models"].keys()))

['fasttext-wiki-news-subwords-300', 'conceptnet-numberbatch-17-06-300', 'word2vec-ruscorpora-300', 'word2vec-google-news-300', 'glove-wiki-gigaword-50', 'glove-wiki-gigaword-100', 'glove-wiki-gigaword-200', 'glove-wiki-gigaword-300', 'glove-twitter-25', 'glove-twitter-50', 'glove-twitter-100', 'glove-twitter-200', '__testing_word2vec-matrix-synopsis']


In [5]:
# This will take a while to run when you run it for the first time.
import gensim.downloader as api

glove_wiki_vectors = api.load("glove-wiki-gigaword-100")

In [6]:
len(glove_wiki_vectors)

400000

There are 400,000 word vectors in this pre-trained model. 

Now that we have GloVe Wiki vectors loaded in `glove_wiki_vectors`, let's explore the embeddings. 

<br><br>

<!-- BEGIN QUESTION -->

### 1.1 Word similarity using pre-trained embeddings
rubric={points:2}

**Your tasks:**

- Come up with a list of 4 words of your choice and find similar words to these words in `glove_wiki_vectors` embeddings. 

<div class="alert alert-warning">

Solution_1.1
    
</div>

In [7]:
glove_wiki_vectors.most_similar("mom")

[('dad', 0.8626731634140015),
 ('grandmother', 0.7560926675796509),
 ('mother', 0.7393314242362976),
 ('aunt', 0.735775887966156),
 ('kids', 0.7256911993026733),
 ('kid', 0.7226841449737549),
 ('daddy', 0.722258448600769),
 ('girl', 0.7197164297103882),
 ('baby', 0.699421226978302),
 ('my', 0.6959770321846008)]

In [8]:
glove_wiki_vectors.most_similar("tree")

[('trees', 0.8510393500328064),
 ('grass', 0.6922113299369812),
 ('pine', 0.6856961846351624),
 ('oak', 0.6387961506843567),
 ('leaf', 0.6364573836326599),
 ('bushes', 0.6336526870727539),
 ('flower', 0.6273626089096069),
 ('green', 0.6112821102142334),
 ('garden', 0.603394091129303),
 ('bark', 0.6031671762466431)]

In [9]:
glove_wiki_vectors.most_similar("house")

[('office', 0.7581615447998047),
 ('senate', 0.7204986810684204),
 ('room', 0.7149738669395447),
 ('houses', 0.6888046264648438),
 ('capitol', 0.6851760149002075),
 ('building', 0.684728741645813),
 ('home', 0.672031044960022),
 ('clinton', 0.6707026958465576),
 ('congressional', 0.669257640838623),
 ('mansion', 0.665092408657074)]

In [10]:
glove_wiki_vectors.most_similar("buy")

[('sell', 0.9384769201278687),
 ('purchase', 0.8328440189361572),
 ('bought', 0.776477575302124),
 ('buying', 0.773395836353302),
 ('acquire', 0.7605097889900208),
 ('buys', 0.7470086812973022),
 ('selling', 0.7394707202911377),
 ('sale', 0.7385357022285461),
 ('offer', 0.738306999206543),
 ('offering', 0.7315958142280579)]

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.2 Word similarity using pre-trained embeddings
rubric={points:2}

**Your tasks:**
1. Calculate cosine similarity for the following word pairs (`word_pairs`) using the [`similarity`](https://radimrehurek.com/gensim/models/keyedvectors.html?highlight=similarity#gensim.models.keyedvectors.KeyedVectors.similarity) method of `glove_wiki_vectors`.

In [11]:
word_pairs = [
    ("coast", "shore"),
    ("clothes", "closet"),
    ("old", "new"),
    ("smart", "intelligent"),
    ("dog", "cat"),
    ("tree", "lawyer"),
]


In [12]:
#Adapted from lecture 18

for pair in word_pairs:
    print(
        "The similarity between %s and %s is %0.3f"
        % (pair[0], pair[1], glove_wiki_vectors.similarity(pair[0], pair[1]))
    )

The similarity between coast and shore is 0.700
The similarity between clothes and closet is 0.546
The similarity between old and new is 0.643
The similarity between smart and intelligent is 0.755
The similarity between dog and cat is 0.880
The similarity between tree and lawyer is 0.077


<div class="alert alert-warning">

Solution_1.2
    
</div>

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.3 Stereotypes and biases in embeddings
rubric={points:6}

Word vectors contain lots of useful information. But they also contain stereotypes and biases of the texts they were trained on. In the lecture, we saw an example of gender bias in Google News word embeddings. Here we are using pre-trained embeddings trained on Wikipedia data. 

**Your tasks:**

1. Explore whether there are any worrisome biases or stereotypes present in these embeddings or not by trying out at least 4 examples. You can use the following two methods or other methods of your choice to explore what kind of stereotypes and biases are encoded in these embeddings. 
    - the `analogy` function below which gives word analogies (an example shown below)
    - [similarity](https://radimrehurek.com/gensim/models/keyedvectors.html?highlight=similarity#gensim.models.keyedvectors.KeyedVectors.similarity) or [distance](https://radimrehurek.com/gensim/models/keyedvectors.html?highlight=distance#gensim.models.keyedvectors.KeyedVectors.distances) methods (an example is shown below)   
2. Discuss your observations.

> Note that most of the recent embeddings are de-biased. But you might still observe some biases in them. Also, not all stereotypes present in pre-trained embeddings are necessarily bad. But you should be aware of them when you use them in your models. 

An example of using analogy to explore biases and stereotypes.  

In [13]:
def analogy(word1, word2, word3, model=glove_wiki_vectors):
    """
    Returns analogy word using the given model.

    Parameters
    --------------
    word1 : (str)
        word1 in the analogy relation
    word2 : (str)
        word2 in the analogy relation
    word3 : (str)
        word3 in the analogy relation
    model :
        word embedding model

    Returns
    ---------------
        pd.dataframe
    """
    print("%s : %s :: %s : ?" % (word1, word2, word3))
    sim_words = model.most_similar(positive=[word3, word2], negative=[word1])
    return pd.DataFrame(sim_words, columns=["Analogy word", "Score"])

In [14]:
analogy("man", "doctor", "woman")

man : doctor :: woman : ?


Analogy word     Score
0        nurse  0.773523
1    physician  0.718943
2      doctors  0.682433
3      patient  0.675068
4      dentist  0.672603
5     pregnant  0.664246
6      medical  0.652045
7      nursing  0.645348
8       mother  0.639333
9     hospital  0.638750

An example of using similarity between words to explore biases and stereotypes.  

In [15]:
glove_wiki_vectors.similarity("white", "rich")

0.44723597

In [16]:
glove_wiki_vectors.similarity("black", "rich")

0.51745194

<div class="alert alert-warning">

Solution_1.3
    
</div>

_Type your answer here, replacing this text._

In [17]:
print("Example 1")
print(glove_wiki_vectors.similarity("asian", "disease"))
print(glove_wiki_vectors.similarity("white", "disease"))
print(glove_wiki_vectors.similarity("asian", "pandemic"))
print(glove_wiki_vectors.similarity("white", "pandemic"))

Example 1
0.32689855
0.23972392
0.24521746
0.017326914


In [18]:
print("Example 2")
print(glove_wiki_vectors.similarity("women", "career"))
print(glove_wiki_vectors.similarity("men", "career"))
print(glove_wiki_vectors.similarity("women", "successful"))
print(glove_wiki_vectors.similarity("men", "successful"))

Example 2
0.40829042
0.42878062
0.45438394
0.4322261


In [19]:
print("Example 3")
print(glove_wiki_vectors.similarity("boy", "pink"))
print(glove_wiki_vectors.similarity("girl", "pink"))
print(glove_wiki_vectors.similarity("boy", "blue"))
print(glove_wiki_vectors.similarity("girl", "blue"))

Example 3
0.41258276
0.43739998
0.46894285
0.43496475


In [20]:
print("Example 4")
print(glove_wiki_vectors.similarity("boy", "action"))
print(glove_wiki_vectors.similarity("girl", "action"))
print(glove_wiki_vectors.similarity("boy", "romance"))
print(glove_wiki_vectors.similarity("girl", "romance"))

Example 4
0.35418442
0.33710176
0.40344507
0.5121082


With the method used above, the higher the score, the higher the similarity between words. In the few examples tested above, we can observe some bias. 

1- With the Covid-19 pandemic, the stereotype of asians carrying diseases has gained strength. In the results above, the word "asian" has a significant similarity to "disease" than the word "white" does. This is further supported by an extreme difference in similarity between "white" and "asian" when associated with "pandemic". This is an indication of bias in the embedding. 

2- Another common stereotype is that men have more successful careers than women. From the results above, that bias does not seem to be having a significant impact on the embeddings. We see a higher similarity of "men" with "career" but also of "women" with "successful". The differences in similarity scores are also not large. 

3- A stereotype related to gender roles is the association with colours. Here we see that the results are consistent with this bias, with the similarity of "boy" being higher than "girl" when compared to "blue" but smaller when compared to "pink". Though the differences here are small, they could indicate bias. 

4- Example four is another exploration of gender bias, this time with themes. Similarly to the colours, we have results consistent with the presence of bias - "action" is more similar to "boy" and "romance" to "girl". However, the differences are again significant but not very large. 

<!-- BEGIN QUESTION -->

### 1.4 Discussion
rubric={points:4}

**Your tasks:**
1. Based on your exploration above, comment on the overall quality of these pre-trained embeddings. 
2. In the lecture, we saw that our pre-trained word embedding model output an analogy that reinforced a gender stereotype. Give an example of how using such a model could cause harm in the real world.

<div class="alert alert-warning">

Solution_1.4
    
</div>

The exploration above is not enough to properly evaluate the quality of these pre-trained embeddings. However, it does raise a concern about the presence of bias. The extreme case of the similarity between "asian" and "pandemic" being 14 times higher than between "white" and "pandemic" is a clear demonstration of this bias. Gender stereotype also appears to have a moderate but consistent impact on the embeddings. The presence of such biases in pre-trained embeddings is concerning. Pre-trained embeddings are very widely used in NLP, which in turn has become part of all of our lives (in search engines, email, social media, etc.). When stereotypes are part of pre-trained embeddings, these same stereotypes will become a part of the systems that are widely used in society. As a direct result, the reinforcement of these stereotypes is strengthened. A possible example of this is when you google "girl colour" and click on images, the vast majority of results are different shades of pink and purple. At the same time when you google "boy colour", almost all results are blue, green, and red, and almost none are pink and purple. When systems like this reinforce the stereotypes, parents will be more likely to paint their child's room, give them clothes or toys of that colour, etc. This continues to reinforce the stereotype and makes the cycle very hard to break. As associations like this reinforce other deeper stereotypes of gender roles and places in society, they will be strengthened as well, making it harder for society to evolve past this.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.5 Classification with pre-trained embeddings 
rubric={points:8}

In lecture 18, we saw that you can conveniently get word vectors with `spaCy` with `en_core_web_md` model. In this exercise, you'll use word embeddings in multi-class text classification task. We will use [HappyDB](https://www.kaggle.com/ritresearch/happydb) corpus which contains about 100,000 happy moments classified into 7 categories: *affection, exercise, bonding, nature, leisure, achievement, enjoy_the_moment*. The data was crowd-sourced via [Amazon Mechanical Turk](https://www.mturk.com/). The ground truth label is not available for all examples, and in this lab, we'll only use the examples where ground truth is available (~15,000 examples). 

- Download the data from [here](https://www.kaggle.com/ritresearch/happydb).
- Unzip the file and copy it under data/ directory in this homework directory.

The code below reads the data CSV (assuming that it's present in the current directory as *cleaned_hm.csv*),  cleans it up a bit, and splits it into train and test splits. 

**Your tasks:**

1. Train logistic regression with bag-of-words features (`CountVectorizer`) and show [classification report](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html) on the test set. 
2. Train logistic regression with average embedding representation extracted using spaCy and show [classification report](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html) on the test set. (You can refer to lecture 18 notes for this.)  

In [21]:
df = pd.read_csv("data/cleaned_hm.csv", index_col=0)
sample_df = df.dropna()
sample_df.head()

wid reflection_period  \
hmid                            
27676   206               24h   
27678    45               24h   
27697   498               24h   
27705  5732               24h   
27715  2272               24h   

                                             original_hm  \
hmid                                                       
27676  We had a serious talk with some friends of our...   
27678                            I meditated last night.   
27697  My grandmother start to walk from the bed afte...   
27705  I picked my daughter up from the airport and w...   
27715        when i received flowers from my best friend   

                                              cleaned_hm  modified  \
hmid                                                                 
27676  We had a serious talk with some friends of our...      True   
27678                            I meditated last night.      True   
27697  My grandmother start to walk from the bed afte...      True   
27705  I picked my daughter up from the airport and w...      True   
27715        when i received flowers from my best friend      True   

       num_sentence ground_truth_category predicted_category  
hmid                                                          
27676             2               bonding            bonding  
27678             1               leisure            leisure  
27697             1             affection          affection  
27705             1               bonding          affection  
27715             1               bonding            bonding

In [22]:
sample_df = sample_df.rename(
    columns={"cleaned_hm": "moment", "ground_truth_category": "target"}
)

In [23]:
train_df, test_df = train_test_split(sample_df, test_size=0.3, random_state=123)
X_train, y_train = train_df["moment"], train_df["target"]
X_test, y_test = test_df["moment"], test_df["target"]

You need `spacy` to run the code below. If it's not in your course conda environment, you need to install it. 

> conda install -c conda-forge spacy

You also need to download the following language model. 

> python -m spacy download en_core_web_md

In [25]:
import spacy

nlp = spacy.load("en_core_web_md")

<div class="alert alert-warning">

Solution_1.5
    
</div>

_Type your answer here, replacing this text._

In [26]:
# Train logistic regression with bag-of-words features
pipe = make_pipeline(
    CountVectorizer(stop_words="english"), LogisticRegression(max_iter=1000)
)

pipe.fit(X_train, y_train);

Ellipsis

In [27]:
# classification report
from sklearn.metrics import classification_report

y_predict = pipe.fit(X_train, y_train).predict(X_test)
target_names = ['affection', 'exercise', 'bonding', 'nature', 'leisure', 'achievement', 'enjoy_the_moment']

print(classification_report(y_test, y_predict, target_names=target_names))

Ellipsis

In [28]:
# Train logistic regression with average embedding representation extracted using spaCy
X_train_embeddings = pd.DataFrame([text.vector for text in nlp.pipe(X_train)])
X_test_embeddings = pd.DataFrame([text.vector for text in nlp.pipe(X_test)])

Ellipsis

In [29]:
lgr = LogisticRegression(max_iter=8000)
lgr.fit(X_train_embeddings, y_train)

Ellipsis

In [30]:
# classification report
from sklearn.metrics import classification_report

y_predict = lgr.fit(X_train_embeddings, y_train).predict(X_test_embeddings)
target_names = ['affection', 'exercise', 'bonding', 'nature', 'leisure', 'achievement', 'enjoy_the_moment']

print(classification_report(y_test, y_predict, target_names=target_names))

Ellipsis

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.6 Discussion
rubric={points:6}

**Your tasks:**
1. Briefly explain the difference between using `CountVectorizer` vs. average-embedding approach for text classification.  
2. Which representation among these two would be more interpretable? Why?   
3. Are we using any transfer learning here? If yes, are you observing any benefits of transfer learning? Briefly discuss. 

<div class="alert alert-warning">

Solution_1.6
    
</div>

_1. CountVectorizer counts the times words appear, average-embedding creates a numerical representation of words and applies a weighting to each word. Average-embedding is more useful when dealing with a big dataset._

_2. CountVectorizer is more easier to understand and more interpretable. We can easily figure out the appearance times of important words in examples and match them with different categories, then find the rules and build the model._

_3. Yes, we are using transfer learning here. spaCy supports transfer learning. Transfer learning can save training time by reusing a model developed for a dataset as the beginning point of another dataset. It refers to techniques such as word embeddings pretraining. Pretrained word embeddings learn in one task and use the embeddings to solve another similar task._

<!-- END QUESTION -->

<br><br><br><br>

## Exercise 2: Topic modeling 

The overarching goal of topic modeling is understanding high-level themes in a large collection of texts in an unsupervised way. 

In this exercise you will explore topics in a subset of `scikit-learn`'s [20 newsgroups text dataset](https://scikit-learn.org/0.19/datasets/twenty_newsgroups.html) using `scikit-learn`'s `LatentDirichletAllocation` (LDA) model. 

Usually, topic modeling is used for discovering abstract "topics" that occur in a collection of documents when you do not know the actual topics present in the documents. But 20 newsgroups text dataset is labeled with categories (e.g., sports, hardware, religion), and you will be able to cross-check the topics discovered by your model with these available topics. 

The starter code below loads the train and test portion of the data and convert the train portion into a pandas DataFrame. For speed, we will only consider documents with the following 8 categories. 

In [31]:
from sklearn.datasets import fetch_20newsgroups

In [32]:
cats = [
    "rec.sport.hockey",
    "rec.sport.baseball",
    "soc.religion.christian",
    "alt.atheism",
    "comp.graphics",
    "comp.windows.x",
    "talk.politics.mideast",
    "talk.politics.guns",
]  # We'll only consider these categories out of 20 categories for speed.

newsgroups_train = fetch_20newsgroups(
    subset="train", remove=("headers", "footers", "quotes"), categories=cats
)
X_news_train, y_news_train = newsgroups_train.data, newsgroups_train.target
df = pd.DataFrame(X_news_train, columns=["text"])
df["target"] = y_news_train
df["target_name"] = [
    newsgroups_train.target_names[target] for target in newsgroups_train.target
]
df

text  target  \
0     You know, I was reading 18 U.S.C. 922 and some...       6   
1     \n\n\nIt's not a bad question: I don't have an...       1   
2     \nActuallay I don't, but on the other hand I d...       1   
3     The following problem is really bugging me,\na...       2   
4     \n\n  This is the latest from UPI \n\n     For...       7   
...                                                 ...     ...   
4558  Hi Everyone ::\n\nI am  looking for  some soft...       1   
4559  Archive-name: x-faq/part3\nLast-modified: 1993...       2   
4560  \nThat's nice, but it doesn't answer the quest...       6   
4561  Hi,\n     I just got myself a Gateway 4DX-33V ...       2   
4562  \n\n[h] \tThe Armenians in Nagarno-Karabagh ar...       7   

                target_name  
0        talk.politics.guns  
1             comp.graphics  
2             comp.graphics  
3            comp.windows.x  
4     talk.politics.mideast  
...                     ...  
4558          comp.graphics  
4559         comp.windows.x  
4560     talk.politics.guns  
4561         comp.windows.x  
4562  talk.politics.mideast  

[4563 rows x 3 columns]

In [33]:
newsgroups_train.target_names

['alt.atheism',
 'comp.graphics',
 'comp.windows.x',
 'rec.sport.baseball',
 'rec.sport.hockey',
 'soc.religion.christian',
 'talk.politics.guns',
 'talk.politics.mideast']

<br><br>

<!-- BEGIN QUESTION -->

### 2.1 Preprocessing using [spaCy](https://spacy.io/)
rubric={points:8}

In this exercise you'll prepare the data for topic modeling using [spaCy](https://spacy.io/). Preprocessing is a crucial step before training an LDA model and it markedly affects topic modeling results. So let's carry out preprocessing. 

**Your tasks:** 

Write code to carry out preprocessing of the "text" column in the dataframe above and store the preprocessed text in a new column called "text_pp" in the dataframe. 


Note that there is no such thing as "perfect" preprocessing. You'll have to make your own judgments and decisions on which tokens are more informative and which ones are less informative for the given task. Some common text preprocessing steps for topic modeling are: 
- getting rid of slashes or other weird characters
- sentence segmentation and tokenization      
- getting rid of urls and email addresses
- getting rid of other fairly unique tokens which are not going to help us in topic modeling  
- excluding stopwords and punctuation 
- lemmatization        

You might have to go back and forth between the preprocessing and topic modeling and interpretation steps in the next exercises. 

> Check out [these available attributes](https://spacy.io/api/token#attributes) for `token` in spaCy which might help you with preprocessing. 

> You can also get rid of words with specific POS tags. [Here](https://spacy.io/api/annotation/#pos-en) is the list of part-of-speech tags used in spaCy. 

> Note that preprocessing the corpus might take some time. So here are a couple of suggestions: 1) During the debugging phase, work on a smaller subset of the data. 2) Once you're done with the preprocessing part, you might want to save the preprocessed data so that you don't run the preprocessing part every time you run the notebook. 


<div class="alert alert-warning">

Solution_2_1
    
</div>

_Type your answer here, replacing this text._

In [44]:
nlp2 = spacy.load("en_core_web_md")

In [78]:
def preprocess(
    doc,
    min_token_len=2,
    irrelevant_pos=["ADV", "ADJ", "CONJ","INTJ", "SCONJ", "PRON", "CCONJ", "PUNCT", "PART", "DET", "ADP", "SPACE", "SYM", "NUM", "X", "XX"],
):
    """
    Given text, min_token_len, and irrelevant_pos carry out preprocessing of the text
    and return a preprocessed string.

    Parameters
    -------------
    doc : (spaCy doc object)
        the spacy doc object of the text
    min_token_len : (int)
        min_token_length required
    irrelevant_pos : (list)
        a list of irrelevant pos tags

    Returns
    -------------
    (str) the preprocessed text
    """

    clean_text = []

    for token in doc:
        if (
            token.is_stop == False  # Check if it's not a stopword
            and len(token) > min_token_len  # Check if the word meets minimum threshold
            and token.pos_ not in irrelevant_pos
        ):  # Check if the POS is in the acceptable POS tags
            lemma = token.lemma_  # Take the lemma of the word
            clean_text.append(lemma.lower())
    return " ".join(clean_text)


In [79]:
df["text_pp"] = [preprocess(text) for text in nlp2.pipe(df["text"])]

In [80]:
df["text_pp"]

0       know read u.s.c. sence wonder help u.s.c. prov...
1       question ref list algorithm think bit point fi...
2       actuallay hand support idea have newsgroup asp...
3       problem bug appreciate help create window chil...
4       upi foreign ministry spokesman ferhat ataman t...
                              ...                        
4558    look software call shadow know raytracer visua...
4559    archive part3 modify subject patch x11r5 relea...
4560    answer question difference feds mandate litera...
4561    get gateway 4dx-33v try configure x11r5 need x...
4562    armenians nagarno karabagh defend right homela...
Name: text_pp, Length: 4563, dtype: object

In [81]:
df

text  target  \
0     You know, I was reading 18 U.S.C. 922 and some...       6   
1     \n\n\nIt's not a bad question: I don't have an...       1   
2     \nActuallay I don't, but on the other hand I d...       1   
3     The following problem is really bugging me,\na...       2   
4     \n\n  This is the latest from UPI \n\n     For...       7   
...                                                 ...     ...   
4558  Hi Everyone ::\n\nI am  looking for  some soft...       1   
4559  Archive-name: x-faq/part3\nLast-modified: 1993...       2   
4560  \nThat's nice, but it doesn't answer the quest...       6   
4561  Hi,\n     I just got myself a Gateway 4DX-33V ...       2   
4562  \n\n[h] \tThe Armenians in Nagarno-Karabagh ar...       7   

                target_name                                            text_pp  
0        talk.politics.guns  know read u.s.c. sence wonder help u.s.c. prov...  
1             comp.graphics  question ref list algorithm think bit point fi...  
2             comp.graphics  actuallay hand support idea have newsgroup asp...  
3            comp.windows.x  problem bug appreciate help create window chil...  
4     talk.politics.mideast  upi foreign ministry spokesman ferhat ataman t...  
...                     ...                                                ...  
4558          comp.graphics  look software call shadow know raytracer visua...  
4559         comp.windows.x  archive part3 modify subject patch x11r5 relea...  
4560     talk.politics.guns  answer question difference feds mandate litera...  
4561         comp.windows.x  get gateway 4dx-33v try configure x11r5 need x...  
4562  talk.politics.mideast  armenians nagarno karabagh defend right homela...  

[4563 rows x 4 columns]

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.2 Justification
rubric={points:2}

**Your tasks:**

Outline the preprocessing steps you carried out in the previous exercise and provide a brief justification for these steps. 

<div class="alert alert-warning">

Solution_2_2
    
</div>

_Word tokenization: used spacy to process text into tokens._

_Added POS tags to remove adverbs, adjectives, nums, spaces, symbols, etc. This is to remove irrelevant words in the text._

_Lemmatization: takes lemma of the words, in order to remove unnecessary grammar, since we only want to keep the simplest form_

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.3 Build a topic model using sklearn's LatentDirichletAllocation
rubric={points:4}

**Your tasks:**
1. Create a topic model on the preprocessed data using [sklearn's `LatentDirichletAllocation`](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.LatentDirichletAllocation.html). Pick a reasonable number for `n_components`, i.e., number of topics and briefly justify your choice.

<div class="alert alert-warning">

Solution_2_3
    
</div>

Code adapted from lecture

In [84]:
from sklearn.feature_extraction.text import CountVectorizer

vec = CountVectorizer(stop_words='english')
X = vec.fit_transform(df["text_pp"])

In [85]:
from sklearn.decomposition import LatentDirichletAllocation

n_topics = 3
lda = LatentDirichletAllocation(
    n_components=n_topics, learning_method="batch", max_iter=10, random_state=0
)
document_topics = lda.fit_transform(X)

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.4 Exploring word topic association
rubric={points:5}

**Your tasks:**
1. Show top 10 words for each of your topics and suggest labels for each of the topics (similar to how we came up with labels "health and nutrition", "fashion", and "machine learning" in the toy example we saw in class). 

> If your topics do not make much sense, you might have to go back to preprocessing in Exercise 2.1, improve it, and train your LDA model again. 

<div class="alert alert-warning">

Solution_2_4
    
</div>

In [86]:
sorting = np.argsort(lda.components_, axis=1)[:, ::-1]
feature_names = np.array(vec.get_feature_names_out())

In [87]:
import mglearn

mglearn.tools.print_topics(
    topics=range(3),
    feature_names=feature_names,
    sorting=sorting,
    topics_per_chunk=5,
    n_words=10,
)

topic 0       topic 1       topic 2       
--------      --------      --------      
file          people        game          
program       god           team          
use           say           year          
image         know          play          
window        think         player        
include       time          win           
server        come          season        
edu           gun           entry         
run           believe       think         
work          thing         hockey        




<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.5 Exploring document topic association
rubric={points:5}

**Your tasks:**
1. Show the document topic assignment of the first five documents from `df`. 
2. Comment on the document topic assignment of the model. 

<div class="alert alert-warning">

Solution_2_5
    
</div>

In [ ]:
df.iloc[0]['text']

In [ ]:
document_topics[0]

The first document is made up of
- ~98.46% topic 1
- ~0.79% topic 0
- ~0.75% topic 2

In [ ]:
df.iloc[1]['text']

In [ ]:
document_topics[1]

The second document is made up of
- ~98.59% topic 0
- ~0.73$% topic 1
- ~0.68% topic 2

In [ ]:
df.iloc[2]['text']

In [ ]:
document_topics[2]

The third document is made up of
- ~52.79% topic 1
- ~46.51% topic 0
- ~0.69% topic 2

In [ ]:
df.iloc[3]['text']

In [ ]:
df.iloc[3]['text']

The fourth document is made up of
- ~91.46% topic 0
- ~7.17% topic 1
- ~1.37% topic 2

In [ ]:
df.iloc[4]['text']

In [ ]:
document_topics[4]

The fifth document is made up of
- ~98.10% topic 1
- ~0.96% topic 0
- ~0.94% topic 2

Comments: Topic modeling is an unsupervised machine learning that analyzes text data to determine cluster words for a set of documents. It doesn't need predefined tags or pretrained data, it classifies based on the input examples. The classification of the first five documents except the third one are very clear with more than 90%  of the classified topic.

<!-- END QUESTION -->

<br><br><br><br>

<!-- BEGIN QUESTION -->

## Exercise 3: Short answer questions 
<hr>

rubric={points:6}

1. In lecture 19, we talked about multi-class classification. Comment on how each model in the list below might be handling multiclass classification. Check `scikit-learn` documentation for each of these models when you answer this question.  
    - Decision Tree
    - KNN
    - Random Forest    
    - Logistic Regression
    - SVM RBF
2. What is transfer learning in natural language processing or computer vision? Briefly explain.     
3. In Lecture 19 we briefly discussed how neural networks are sort of like `Pipeline`s, in the sense that they involve multiple sequential transformations of the data, finally resulting in the prediction. Why was this property useful when it came to transfer learning?

<div class="alert alert-warning">

Solution_3
    
</div>

1.
* Decision Tree can naturally handle multi-class problems by default built-in. We can just train the model similarly to the binary case where there are only 2 classes, but here our Y train just contains more classes. After training the model on the training data and use it for prediction, decision tree model will predict the class with highest probability, and if there are several classes with the same highest probability then it will take the class with the lowest index.
* KNN: KNN can also handle multi-class classification with the built in model without any further specification. When there is a new example that needs to be predicted, the model will calculate the distance to close by examples and pick the closest K neighbors, then take the class of the majority of the K neighbors to be the prediction; and this process go the same for a binary problem or a multi-class problem.
* Random Forest: As Random Forest is built on multiple Decision Trees, it can also handle multi-class by default. We can specify the parameters as well (class_weight) in case of multi-class just as with a 2 classes problem.
* Logistic Regression: Logistic Regression is able to handle multi-class classification by setting the parameter multi_class to 'ovr' (one vs rest) or 'multinomial'. If 'ovr' is chosen, the model will learn several binary linear models, one for each class.
* SVM RBF: SVM RBF does support multi-class classification using the one vs one approach. However, we can specify if we want to return the decision function as one vs one or one vs rest by specifying the decision_function_shape parameter. However, internally, the model will still follow the one vs one scheme.
(Answers adapted from scikit-learn document)

2. Transfer learning is a technique where we utilize pre-trained models. This is especially helpful in natural language processing or computer vision because to train new models we need a very large dataset, storage space as well as a lot of work and time. Using pre-trained models that have been successful and applied across different problems directly or to build our own new model from that will save time and help with accuracy.

3. Not all pre-trained models work well with our specific tasks, so the fact that the neural network is similar to a Pipeline allows us to make use of the pre-trained models to extract features until the last layer before prediction, but then use our own classifier at the end using the extracted features to make the final prediction. Being similar to a Pipeline is useful for us to adjust and adapt the pre-trained models for our tasks.

<!-- END QUESTION -->

<br><br><br><br>

**PLEASE READ BEFORE YOU SUBMIT:** 

When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order or not starting from "1" will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using Gradescope's drag and drop tool. Check out this [Gradescope Student Guide](https://lthub.ubc.ca/guides/gradescope-student-guide/) if you need help with Gradescope submission. 
4. Make sure that the plots and output are rendered properly in your submitted file. If the .ipynb file is too big and doesn't render on Gradescope, also upload a pdf or html in addition to the .ipynb so that the TAs can view your submission on Gradescope. 

![](img/eva-well-done.png)